### PR TITLE
Issue#681 Automatically convert HTML entities from scrapers back into normal characters

### PR DESCRIPTION
**Pull request by hisayohorie**
this is to fix <a href="https://github.com/openaustralia/planningalerts/issues/681"> issue#681</a>

I added htmlentities gem, and created a method called decode in applications_helper, and used decode in applications/index.html.haml as well as _application.html, and a wrote a test in applications_helper_spec.rb
@equivalentideas @henare

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -1,4 +1,4 @@
-source "https://rubygems.org"
+source "http://rubygems.org"
```
Comment (henare) [inline Gemfile]: This shouldn't be changed to insecure HTTP.

File: Gemfile [hunk 2]
```diff
@@ -78,6 +79,7 @@ group :test do
 end
 
 group :development do
+  gem 'letter_opener'
```
Comment (henare) [inline Gemfile]: There's a few additions to the Gemfile from here down that don't seem to be related to this change so they should not be included in this pull request.

File: app/helpers/applications_helper.rb [hunk 3]
```diff
@@ -24,6 +24,12 @@ def scraped_and_received_text(application)
     text
   end
 
+  def decode(description)
+    coder = HTMLEntities.new
+    first_decoding = coder.decode(description)
```
Comment (henare) [inline app/helpers/applications_helper.rb]: Why are there 2 decode calls?

File: app/helpers/applications_helper.rb [hunk 4]
```diff
@@ -24,6 +24,12 @@ def scraped_and_received_text(application)
     text
   end
 
+  def decode(description)
```
Comment (henare) [inline app/helpers/applications_helper.rb]: We already do a lot of manipulation to the description in `Application#description`. Since this would put this logic in one place I think we should probably put this there too.

File: spec/helpers/applications_helper_spec.rb [hunk 5]
```diff
@@ -85,6 +85,12 @@
     end
   end
 
+  describe "decode" do
+    it "should decode the html entities such as &amp; to &" do
```
Comment (henare) [inline spec/helpers/applications_helper_spec.rb]: For some reason my instincts are telling me we need more tests around the edge cases. My instincts could be wrong :smile: 

Are there situations where we wouldn't want to decode HTML entities? Are the potential security issues we could be introducing to the page?

It might be good to have a look on PlanningAlerts for real-world examples of where this is a problem and specifically test them too. I always think if you've found a problem it's best to add tests for that exact problem, and the other success, failure, and edge cases you can think of.

File: Gemfile (hunk 2 above)
Comment (henare) [inline Gemfile]: It looks like one of these additions has also [broken the build](https://travis-ci.org/openaustralia/planningalerts/builds/223779317) so hopefully fixing this up with fix the build too :sparkles:

File: app/helpers/applications_helper.rb (hunk 3 above)
Comment (hisayohorie) [inline app/helpers/applications_helper.rb]: @henare I had to call it twice to convert some entities. The first call converts &amp; to &, and the second call converts &#8217 to ', for example. If you know of better ways to do it or some insight to start, please let me know.

File: spec/helpers/applications_helper_spec.rb (hunk 5 above)
Comment (hisayohorie) [inline spec/helpers/applications_helper_spec.rb]: @henare hmm I can't think of the case we don't want to decode it or potential security issues...any insights? I agree with looking into PlanningAlerts real world examples. I will look into it this week.

File: Gemfile (hunk 1 above)
Comment (hisayohorie) [inline Gemfile]: true!

File: Gemfile (hunk 2 above)
Comment (hisayohorie) [inline Gemfile]: also true!

File: app/helpers/applications_helper.rb (hunk 3 above)
Comment (henare) [inline app/helpers/applications_helper.rb]: That's really helpful @hisayohorie :+1: I don't know of any better ways off the top of my head but this would be a great place to put a comment to explain why we're calling this twice.

Whenever code is difficult to understand first try to simplify it so it is easy to understand :smile: but if you can't then that's when a comment is the way to go :ok_hand:

File: spec/helpers/applications_helper_spec.rb (hunk 5 above)
Comment (henare) [inline spec/helpers/applications_helper_spec.rb]: >I can't think of the case we don't want to decode it or potential security issues...any insights?

Is there a case where it could potentially inject a `<script>` tag? e.g. by decoding an HTML entity for `<`? I'm honestly not sure if this is a problem here but this area of code, where you're manipulating outside external data and injecting it onto a page, is the place to be careful for this kind of security issue.

Comment (hisayohorie) [inline spec/helpers/applications_helper_spec.rb]: I tried using <script></script> both encoded and decoded, and both cases browser treated it like a text. Can you think of any other ways  of testing it?

Comment (henare) [inline spec/helpers/applications_helper_spec.rb]: Nice :smile: that should be all good then. I guess Rails is still automatically escaping it.

Comment (hisayohorie) [inline spec/helpers/applications_helper_spec.rb]: I moved the decode under application#description and added a comment. still need to look into real life examples and write a test accordingly!

**Comment by hisayohorie** [comment]
@henare Here is the PR for #1145 !

**Comment by handelaar** [comment]
(If we are fixing stray entities in the description field we should do the same to the address field. Plenty of street names contain apostrophes. )

**Comment by mlandauer** [comment]
It pains me to close an open PR without merging it when it's clear how much work has gone into this but unfortunately this has been open for long enough that it would take a bit to get it working against the current master.

Also, I wonder if #681 is suggesting fixing the problem in the right way? Strictly the scraper should be the one that knows how to handle html entities and any other encoding issues because it has the context of the page the data is being scraped from.
